toolchain/docs/parse.md

```diff
@@ -0,0 +1,758 @@
+      {kind: 'ParamList', text: ')', subtree_size: 2},
+        {kind: 'Literal', text: 'f64'},
+      {kind: 'ReturnType', text: '->', subtree_size: 2},
+    {kind: 'FunctionDefinitionStart', text: '{', subtree_size: 7},
+      {kind: 'ReturnStatementStart', text: 'return'},
+      {kind: 'Literal', text: '42'},
+    {kind: 'ReturnStatement', text: ';', subtree_size: 3},
+  {kind: 'FunctionDefinition', text: '}', subtree_size: 11},
+  {kind: 'FileEnd', text: ''},
+]
+```
+
+<!-- prettier-ignore-end -->
+
+In this example, `FileStart`, `FunctionDefinition`, and `FileEnd` are "root"
+nodes for the tree. Function components are children of `FunctionDefinition`.
+
+It's produced in this way because it's an efficient layout to produce with
+vectorized storage, requiring little context to be maintained during parsing.
+Because it's stored in postorder, it's also most efficient to process the parsed
+output in postorder; this affects checking.
+
+The parse tree is printed in postorder by default because it matches how the
+parse tree is expected to be processed within the toolchain , and so can make it
+easier to reason about. However, the `--preorder` flag may be used in contexts
+where a preorder representation would be easier to handle.
+
+## Bracketing inside the tree
+
+The parse tree is designed to be walked in postorder by checking, allowing
+checking to be more efficient. To support this, checking sometimes requires
+context on the meaning of a node when it is encountered.
+
+Each `ParseNodeKind` has either a bracketing node, or a specific child count.
+This helps document and enforce the expected tree structure.
+
+When a bracketing node is indicated, it is the opening bracket: it will always
+be the first child of the parent, and that will be the only time it occurs in
+the parent's children (it may still occur in children of children). When
+checking encounters the opening bracket, this means it can make contextual
+decisions for the later children of the node.
+
+Nodes can also have a specific child count, for example, infix operators always
+have two children: the lhs and rhs expressions. Many nodes have a child count of
+0; this just means they're leaf nodes, and will never have children.
+
+Because the tree structure is always valid, these are treated as contracts. Some
+nodes exist only to be used to construct valid tree structures for invalid
+input, such as `StructFieldUnknown`.
+
+Although each subtree's size is also tracked as part of the node, we're
+currently trying to avoid relying on it and may eliminate it if it turns out to
+be unnecessary and a meaningful cost for the compiler.
+
+## Visual example
+
+To try to explain the transition from code to Parse Tree, consider the
+statement:
+
+```carbon
+var x: i32 = y + 1;
+```
+
+Lexing creates distinct tokens for each syntactic element, which will form the
+basis of the parse tree:
+
+```ascii
++-----+ +---+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+
```

Comment:
```suggestion
Tokens:
+-----+ +---+ +---+ +-----+ +---+ +---+ +---+ +---+ +---+
```

(See comment on line 327 for the rationale for this and the following suggestions)

Reply:
Note, I was omitting labels in later charts because they are not in the toolchain architecture doc.